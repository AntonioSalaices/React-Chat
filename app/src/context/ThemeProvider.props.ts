export interface ThemeProviderProps {
  /**
   * Children node
   */
  children: JSX.Element | JSX.Element[];
}
