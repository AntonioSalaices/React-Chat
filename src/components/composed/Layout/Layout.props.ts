export interface LayoutProps {
  /**
   * Children node
   */
  children?: JSX.Element;
}
