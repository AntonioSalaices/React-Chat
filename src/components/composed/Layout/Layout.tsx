import { useContext, useEffect, useState } from 'react';

import { Section, Header, Footer } from '@Components/Core';

import { Outlet } from 'react-router-dom';
import { AuthModal } from 'screens/auth/auth';
import { AuthContext } from 'auth/AuthContext';

const Layout: React.FC = () => {
  const { user, logout } = useContext(AuthContext);
  const [isShownLoginModal, setIsShownLoginModal] = useState<boolean>(false);

  useEffect(() => {
    if (user) {
      setIsShownLoginModal(false);
    }
  }, [user]);

  return (
    <div>
      <Header
        onClickLoginNavigation={() => setIsShownLoginModal(!isShownLoginModal)}
        logged={!!user}
        onClickLogout={logout}
      />
      <Outlet />
      <Section />
      <Footer tx="footer.message" />
      <AuthModal onClose={() => setIsShownLoginModal(!isShownLoginModal)} isShown={isShownLoginModal} />
    </div>
  );
};

export { Layout };
