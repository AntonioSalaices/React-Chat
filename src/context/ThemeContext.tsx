import { Theme } from '@Constans/Theme';
import React, { createContext, useEffect, useState } from 'react';
import { ThemeProviderProps } from './ThemeProvider.props';

export const ThemeContext = createContext<any>({
  theme: Theme.Light,
  undefined,
});

export const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
  // We can use local storage to save theme choice
  const [theme, setTheme] = useState<string>(Theme.Light);

  useEffect(() => {
    const getThemeUserPreferences = () => {
      const theme = localStorage.getItem(Theme.Theme);
      theme && setTheme(theme);
    };
    getThemeUserPreferences();
  }, []);

  return <ThemeContext.Provider value={{ theme, setTheme }}>{children}</ThemeContext.Provider>;
};
