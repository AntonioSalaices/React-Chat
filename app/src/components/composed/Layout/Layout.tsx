import { useContext } from 'react';

import Header from '@Components/basics/Header/Header';
import Footer from '@Components/basics/Footer';
import Section from '../Section';

import { ThemeContext } from 'context/ThemeContext';
import { Theme } from '@Constans/Theme';
import { LayoutProps } from './Layout.props';

const Layout: React.FC<LayoutProps> = ({ children }) => {
  const { theme, setTheme } = useContext(ThemeContext);

  const handleOnChange = (selectedTheme: Theme) => {
    setTheme(selectedTheme);
    localStorage.setItem(Theme.THEME, selectedTheme);
  };

  return (
    <div className={theme}>
      <Header theme={theme} onChange={handleOnChange} />
      {children}
      <Section />
      <Footer tx="footer.message" />
    </div>
  );
};

export default Layout;
