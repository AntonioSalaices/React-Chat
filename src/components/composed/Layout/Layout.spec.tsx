import { RenderResult, act, fireEvent, render } from '@testing-library/react';
import Layout from './Layout';
import { ThemeProvider } from 'context/ThemeContext';
import { BrowserRouter } from 'react-router-dom';

const ThemeProviderMock = ThemeProvider as jest.MockedFunction<typeof ThemeProvider>;

describe('<Layout>', () => {
  let component: RenderResult;
  jest.spyOn(Object.getPrototypeOf(window.localStorage), 'setItem');
  Object.setPrototypeOf(window.localStorage.setItem, jest.fn());

  beforeEach(() => {
    component = render(
      <ThemeProviderMock>
        <BrowserRouter>
          <Layout />
        </BrowserRouter>
      </ThemeProviderMock>
    );
  });

  test('should render correctly', () => {
    const { container } = component;

    expect(container).toMatchSnapshot();
  });

  test('should fire onChange theme callback and store new theme preference on local storage', () => {
    const { getByTestId } = component;

    act(() => {
      fireEvent.click(getByTestId('btn-change-theme'));
    });

    expect(window.localStorage.setItem).toHaveBeenCalled();
  });
});
