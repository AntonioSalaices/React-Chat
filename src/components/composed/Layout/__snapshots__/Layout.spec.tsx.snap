// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout> should render correctly 1`] = `
<div>
  <div>
    <nav
      class="navbar-primary text-white mb-4 justify-between"
    >
      <div
        class="container"
      >
        <h2
          class="site-title"
        >
          [missing "en-US.header.title" translation]
          <li
            class="nav-icon"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 416H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
              />
            </svg>
          </li>
        </h2>
        <ul
          class="display-f justify-center align-center gap-2 items-list closed"
        >
          <li>
            <svg
              data-testid="desktop"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 640 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M624 416H381.54c-.74 19.81-14.71 32-32.74 32H288c-18.69 0-33.02-17.47-32.77-32H16c-8.8 0-16 7.2-16 16v16c0 35.2 28.8 64 64 64h512c35.2 0 64-28.8 64-64v-16c0-8.8-7.2-16-16-16zM576 48c0-26.4-21.6-48-48-48H112C85.6 0 64 21.6 64 48v336h512V48zm-64 272H128V64h384v256z"
              />
            </svg>
          </li>
          <li>
            <a
              href="/"
            >
              [missing "en-US.header.links.home" translation]
            </a>
          </li>
          <li>
            <a
              href="/library"
            >
              [missing "en-US.header.links.library" translation]
            </a>
          </li>
          <li>
            <a
              href="/chat"
            >
              [missing "en-US.header.links.chat" translation]
            </a>
          </li>
          <li>
            <button
              class="btn-light text-dark font-md"
              data-testid="btn-login-navigate"
            >
              [missing "en-US.header.links.logOut" translation]
            </button>
          </li>
        </ul>
      </div>
    </nav>
    <section
      class="bg-primary-light-6 mt-5 pt-4 pb-4"
      id="about"
    >
      <div
        class="container"
      >
        <h2
          class="mb-2"
        >
          About React
        </h2>
        <p>
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Culpa ipsam animi aliquid sequi fuga, nam nesciunt dolore libero dolorem exercitationem aliquam cupiditate atque illo, quae dicta doloribus et? Ab ipsam inventore quam asperiores, sequi unde tenetur accusamus, distinctio magni necessitatibus quis deserunt id alias. Iste eum ea labore rerum voluptatibus.
        </p>
        <p
          class="mt-1"
        >
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Soluta nam, corrupti dolorum inventore perspiciatis id illum repellat iste amet sapiente ducimus nihil molestias quasi, totam, ratione minima molestiae blanditiis iure consequatur praesentium debitis. Nulla maiores doloremque tempore nobis dolorum amet!
        </p>
      </div>
    </section>
    <footer
      class="bg-secondary-light-7 pt-3 pb-3 mt-5"
    >
      <div
        class="container"
      >
        [missing "en-US.footer.message" translation]
      </div>
    </footer>
  </div>
</div>
`;
